Fixes Bug1097988 dataservices in api

selenamarie opened the pull request:
This patch implements a dataservice app that loads configuration and objects with configman and points two crashstats.models classes (Bugs and SignaturesByBugs) at those objects.

I created a decorator for caching: `@memoize` in `webapp-django/crashstats/crashstats/models.py`.  It's purpose is to support `CACHE_MIDDLEWARE` and `CACHE_MIDDLEWARE_FILES` config options for the django app.  I removed the 'dont_cache' option as it is no longer necessary. The decorator will be used on every method that needs it and left off of the methods that don't. The previous implementation funnelled all RESTful functions through the same `fetch` function that implemented the caching. 

Config will be done in `webapp-django/crashstats/settings/local.py` in a dict whose default will be in the corresponding `-dist` file. The local.py on stage/production will need to be updated before pushing code. The change should not cause a problem with *existing* code at all.

@peterbe r?

selenamarie commented on webapp-django/crashstats/crashstats/models.py:
```diff
@@ -161,7 +161,96 @@ def inner(*args, **kwargs):
+    using CACHE_MIDDLEWARE and/or CACHE_MIDDLEWARE_FILES Django config"""
+
+    @functools.wraps(function)
+    def memoizer(*args, **kwargs):
```
Ah yes, that makes it more clear.

selenamarie commented on webapp-django/crashstats/crashstats/models.py:
```diff
@@ -161,7 +161,96 @@ def inner(*args, **kwargs):
+    @functools.wraps(function)
+    def memoizer(*args, **kwargs):
+
+        def get_cached_result(key):
```
Yes! That was an error in refactoring on my part, rather than intentional. Thanks for catching it.

selenamarie commented on webapp-django/crashstats/crashstats/models.py:
```diff
@@ -161,7 +161,96 @@ def inner(*args, **kwargs):
+                if instance.expect_json:
+                    json.dump(result, open(cache_file, 'w'), indent=2)
+                else:
+                    open(cache_file, 'w').write(result)
```
Oh good idea. Thanks.

selenamarie commented on webapp-django/crashstats/crashstats/models.py:
```diff
@@ -161,7 +161,96 @@ def inner(*args, **kwargs):
+                result = function(*args, **kwargs)
+                refresh_caches(result)
+            return result
+
```
Ahhh yeah, good catch. That is much nicer. Thanks.

selenamarie commented on webapp-django/crashstats/crashstats/models.py:
```diff
@@ -161,7 +161,96 @@ def inner(*args, **kwargs):
+                cache_file = os.path.join(
+                    cache_file,
+                    '%s.dump' % key
+                )
```
It was repeated twice. I thought I was copying existing functionality, but probably missed something.

I like your suggested version much better.

peterbe commented on socorro/external/postgresql/service_base.py:
```diff
@@ -31,6 +31,30 @@ class PostgreSQLWebServiceBase(DataserviceWebServiceBase):
+    required_config.add_option(
+        'required_params',
+        default=('signatures',),
+    )
```
These two options above don't belong in the service_base.py

selenamarie replied:
oh yeah.. sigh.. /me moves them back

peterbe replied:
sorry :)

peterbe commented on webapp-django/crashstats/settings/local.py-dist:
```diff
@@ -168,3 +168,19 @@ SECRET_KEY = 'you must change this'
+)
+
+DATASERVICE_INI = build_path + '/config/dataservice.ini'
+assert os.path.isfile(DATASERVICE_INI), DATASERVICE_INI
```
Let's move all of this to `settings/base.py`.

peterbe replied:
Basically, `settings/local.py` is just for exceptional overrides and things, like passwords, that can never be in `settings/base.py`.

peterbe commented on webapp-django/crashstats/settings/local.py-dist:
```diff
@@ -168,3 +168,19 @@ SECRET_KEY = 'you must change this'
+# this configuration is for our travis environment
+import os
+build_path = os.environ.get(
+    'TRAVIS_BUILD_DIR',
```
We should perhaps move this to one of the tests. Don't we have something specific in the ./scripts/ files that is specific to travis?

peterbe replied:
Maybe not.

peterbe commented on scripts/bootstrap.sh:
```diff
@@ -36,6 +36,15 @@ make
+    fi
+done
+popd
+
```
If we no longer are going to depend on a `config/dataservice.ini` file, can we move this back into `scripts/test.sh`

peterbe commented:
I'm r+ on the code as it appears all my many nits have been addressed (apart from some Bug specific options being service_base.py)

I haven't tested it locally on my laptop yet. I'm going to do that tomorrow.

selenamarie commented:
@peterbe OK I have this working on my dev instance :) 

The instructions I made a while back for testing dataservice are here: https://gist.github.com/selenamarie/fec696bf042164ea123e

selenamarie commented:
@peterbe alright! I have updated base.py and local.py-dist and I updated our stage config with the changes. I had to add a try/except on stage because the code isn't pushed yet.  

Are we good to merge?

peterbe commented on webapp-django/crashstats/settings/local.py-dist:
```diff
@@ -168,3 +168,43 @@ SECRET_KEY = 'you must change this'
+#try:
+#    DATASERVICE_CONFIG_BASE = base.DATASERVICE_CONFIG_BASE
+#except AttributeError:
+#    pass
```
This doesn't make any sense. 
First of all, we would always have a `base.DATASERVICE_CONFIG_BASE` and if we did get an `AttributeError` it wouldn't later be possible to do `DATASERVICE_CONFIG_BASE.update(`

selenamarie replied:
If someone was deploying this change and upgrading from a previous version of socorro, and they pushed their config changes before updating their app -- without this try/except block our config would break their app.

peterbe replied:
Why would why they have a `local.py-dist` that is more up to date than
`settings/base.py`?!


On Mon, Jan 12, 2015 at 2:30 PM, Selena Deckelmann <notifications@github.com
> wrote:

> In webapp-django/crashstats/settings/local.py-dist
> <https://github.com/mozilla/socorro/pull/2506#discussion-diff-22828498>:
>
> > @@ -168,3 +168,43 @@ SECRET_KEY = 'you must change this'
> >  #           'component=General&groups=client-services-security',
> >  #    'label': 'File a bug in bugzilla'
> >  #}
> > +
> > +
> > +# Dataservice API configuration
> > +# Extend dataservices settings from settings/base.py here
> > +# At a minimum, you'll probably want to change db username/password. All
> > +# dataservice objects inherit resource configuration and so can all
> > +# have their database resource configuration set once in 'secrets.postgresql'
> > +# and 'resource.postgresql' keys.
> > +# Here's an example:
> > +#from . import base
> > +#try:
> > +#    DATASERVICE_CONFIG_BASE = base.DATASERVICE_CONFIG_BASE
> > +#except AttributeError:
> > +#    pass
>
> If someone was deploying this change and upgrading from a previous version
> of socorro, and they pushed their config changes before updating their app
> -- without this try/except block our config would break their app.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/socorro/pull/2506/files#r22828498>.
>



-- 
Peter Bengtsson
Mozilla Web Engineering

selenamarie replied:
Haha - well in our case, we needed to push `local.py` out before we pushed code. Is it that terrible? I don't mind removing it... I only put it in there because it caught me by surprise when I tested the change before I committed it to our puppet repo.

peterbe replied:
That's something we should do in **our** local.py then. Also, it needs indentation to work at all. I mean the `DATASERVICE_CONFIG_BASE.update({` line should be between the `try:` and `except`

peterbe commented on webapp-django/crashstats/settings/local.py-dist:
```diff
@@ -168,3 +168,43 @@ SECRET_KEY = 'you must change this'
+#    'secrets': {
+#       'postgresql': {
+#           'database_password': 'newpassword',
+#            'database_username': 'newusername',
```
Can you add the others too. Like `database_name` and `database_hostname`

selenamarie replied:
Yeah, no problem.

peterbe commented:
Two last comments. But if you fix that, r+
